diff --git a/src/Plugin/views/filter/CshsTaxonomyIndex.php b/src/Plugin/views/filter/CshsTaxonomyIndex.php
index 1806892..5ed87e2 100644
--- a/src/Plugin/views/filter/CshsTaxonomyIndex.php
+++ b/src/Plugin/views/filter/CshsTaxonomyIndex.php
@@ -3,9 +3,12 @@
 namespace Drupal\cshs\Plugin\views\filter;
 
 use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Entity\EntityReferenceSelection\SelectionPluginManagerInterface;
 use Drupal\Core\Entity\EntityRepositoryInterface;
 use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Messenger\MessengerInterface;
 use Drupal\cshs\CshsOptionsFromHelper;
+use Drupal\views\Plugin\views\filter\EntityReference;
 use Drupal\views\ViewExecutable;
 use Drupal\views\Plugin\views\display\DisplayPluginBase;
 use Symfony\Component\DependencyInjection\ContainerInterface;
@@ -24,16 +27,19 @@ trait CshsTaxonomyIndex {
     array $configuration,
     $plugin_id,
     $plugin_definition,
+    SelectionPluginManagerInterface $selection_plugin_manager,
     EntityTypeManagerInterface $entity_type_manager,
-    EntityRepositoryInterface $entity_repository
+    EntityRepositoryInterface $entity_repository,
+    MessengerInterface $messenger
   ) {
-    /* @see \Drupal\taxonomy\Plugin\views\filter\TaxonomyIndexTid::__construct() */
+    /* @see Drupal\views\Plugin\views\filter\EntityReference::__construct() */
     parent::__construct(
       $configuration,
       $plugin_id,
       $plugin_definition,
-      $entity_type_manager->getStorage('taxonomy_vocabulary'),
-      $entity_type_manager->getStorage('taxonomy_term')
+      $selection_plugin_manager,
+      $entity_type_manager,
+      $messenger
     );
 
     $this->entityRepository = $entity_repository;
@@ -42,13 +48,15 @@ trait CshsTaxonomyIndex {
   /**
    * {@inheritdoc}
    */
-  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition): EntityReference {
     return new static(
       $configuration,
       $plugin_id,
       $plugin_definition,
+      $container->get('plugin.manager.entity_reference_selection'),
       $container->get('entity_type.manager'),
-      $container->get('entity.repository')
+      $container->get('entity.repository'),
+      $container->get('messenger')
     );
   }
 
@@ -64,7 +72,7 @@ trait CshsTaxonomyIndex {
   /**
    * {@inheritdoc}
    */
-  public function defineOptions() {
+  public function defineOptions() : array {
     $options = parent::defineOptions();
 
     foreach (static::defaultSettings() + ['type' => static::ID] as $option => $value) {
@@ -136,21 +144,21 @@ trait CshsTaxonomyIndex {
    * {@inheritdoc}
    */
   protected function getVocabularyStorage() {
-    return $this->vocabularyStorage;
+    return $this->entityTypeManager->getStorage('taxonomy_vocabulary');
   }
 
   /**
    * {@inheritdoc}
    */
   protected function getTermStorage() {
-    return $this->termStorage;
+    return $this->entityTypeManager->getStorage('taxonomy_term');
   }
 
   /**
    * {@inheritdoc}
    */
   public function getVocabulary() {
-    return $this->vocabularyStorage->load($this->options['vid']);
+    return $this->getVocabularyStorage()->load($this->options['vid']);
   }
 
 }
